### PR TITLE
Starting type translation and stage2 pass

### PR DESCRIPTION
**Pull request by mariari**

### CODE REVIEW

File: src/Juvix/FrontendDesugar/RemoveDo/Transform.hs [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+
+import qualified Data.List.NonEmpty as NonEmpty
+import qualified Juvix.FrontendDesugar.RemoveDo.Types as New
+import qualified Juvix.FrontendDesugar.RemovePunned.Types as Old
+import Juvix.Library
+
+-- The actual transform we are doing
+
+transformDo :: Old.Do -> New.Expression
+transformDo (Old.Do'' body) = foldr f (transformLast last) body
```
Comment (cwgoes) [inline src/Juvix/FrontendDesugar/RemoveDo/Transform.hs:11]: ... we need to also solve the problem of lifting application over an error monad

Comment (cwgoes) [reply src/Juvix/FrontendDesugar/RemoveDo/Transform.hs:11]: ... and a state monad

Comment (cwgoes) [reply src/Juvix/FrontendDesugar/RemoveDo/Transform.hs:11]: maybe we should use `MonadIdentity` ( =/ )

File: src/Juvix/Frontend/Types/Base.hs [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+      deriving (Show, Generic, NFData)
+
+    data Product
+      = Record !Record
+      | Arrow Expression
+      | ADTLike [Expression]
+      deriving (Show, Generic, NFData)
+
+    data Record
+      = Record''
```
Comment (andy-morris) [inline src/Juvix/Frontend/Types/Base.hs:135]: I wonder if we should maybe come up with a different naming scheme for these things than variously having a single/double `'`?

Maybe like `RecordInfo` or something, but I'm not that fussed about what exactly if someone has a better suggestion.

Comment (mariari) [reply src/Juvix/Frontend/Types/Base.hs:135]: Yes  I would like to do that at some point

**Pull request merged**